### PR TITLE
Run expensive per-FOV pixel clustering processes in parallel

### LINKED ISSUE #532: Parallelize pixel clustering
**Is your feature request related to a problem? Please describe.**
Currently, most of the runtime in the pixel clustering pipeline is for preprocessing, assigning clusters after training, and re-assigning clusters after manual re-clustering.

**Describe the solution you'd like**
Parallelize these steps (probably multiprocessing). For these steps, each fov is independent of each other.

### PR DESCRIPTION
**Pull request by alex-l-kong**
**What is the purpose of this PR?**

Closes #532. There are many processes for pixel clustering that run on a one-per-FOV basis that can be optimized with parallelization.

**How did you implement your changes**

We target four functions:

1. `create_pixel_matrix`
2. `run_pixel_som`
3. `pixel_consensus_cluster`
4. `apply_pixel_meta_cluster_remapping`

1 and 4 are done in Python and rely on the `multiprocessing` module to generate separate processes in parallel. 2 and 3 are done in R, which provides a much cleaner solution using `parallel`, `doParallel`, and `foreach`.

### CODE REVIEW

**Comment by alex-l-kong** [comment]
@ngreenwald @cliu72 I haven't worked out everything yet, but if you want to test a rough draft of the parallelization, you can run it on this PR.

Just one note: I haven't gotten to updating the Dockerhub image yet; it'll need to install certain libraries needed for multiprocessing in R. At the top of `example_pixel_clustering.ipynb` add the following cell for now:

```
!R -e "install.packages('doParallel')"
!R -e "install.packages('foreach')"
```

**Comment by alex-l-kong** [comment]
Bit of a holdup with `pytest`, which isn't playing too nicely with the `multiprocessing` module as it stands. Working on finding workarounds.

**Comment by alex-l-kong** [comment]
@ngreenwald in `test_create_pixel_matrix`, does the following test check anything that the parametrized tests don't (I have a test for if `channel_norm.feather` exists and if it doesn't)?

```
        # check that correct values are passed to helper function
        mocker.patch('ark.phenotyping.som_utils.create_fov_pixel_data',
                     mocked_create_fov_pixel_data)

        if sub_dir is None:
            sub_dir = ''

        # create fake data where all channels in each fov are the same
        new_tiff_dir = os.path.join(temp_dir, 'new_tiff_dir')
        os.makedirs(new_tiff_dir)
        for fov in fovs:
            img = np.random.rand(100).reshape((10, 10))
            fov_dir = os.path.join(new_tiff_dir, fov, sub_dir)
            os.makedirs(fov_dir)
            for chan in chans:
                io.imsave(os.path.join(fov_dir, chan + '.tiff'), img)

        # create normalization file
        data_dir = os.path.join(temp_dir, 'pixel_mat_data')

        # generate the data
        mults = [1 * (1 / 2) ** i for i in range(len(chans))]

        sample_channel_norm_df = pd.DataFrame({'channel': chans,
                                               'norm_val': mults})
        feather.write_dataframe(sample_channel_norm_df,
                                os.path.join(data_dir, 'channel_norm.feather'),
                                compression='uncompressed')

        som_utils.create_pixel_matrix(fovs=fovs,
                                      channels=chans,
                                      base_dir=temp_dir,
                                      tiff_dir=new_tiff_dir,
                                      img_sub_folder=sub_dir,
                                      seg_dir=seg_dir,
                                      dtype='float32',
                                      batch_size=5)
```

**Comment by ngreenwald** [comment]
Yes, this tests that the right values are passed, whereas the other tests check that any output is created

**Comment by alex-l-kong** [comment]
@ngreenwald oh I see now, thanks for clarifying.

**Comment by alex-l-kong** [comment]
Ok parallelization should be finalized, last step is to verify the outputs match up. In the meantime, @cliu72 feel free to test it out yourself.

**Comment by alex-l-kong** [comment]
> Ok parallelization should be finalized, last step is to verify the outputs match up. In the meantime, @cliu72 feel free to test it out yourself.

So unfortunately, `multiprocessing` means the random seed will not take the same random indices per FOV as the current implementation in `master`. The experiment should be replicable, but it will produce a different output because of this.

**Comment by alex-l-kong** [comment]
New Docker image has been pushed with all the dependencies downloaded properly.

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks good. Did you already test out the R code? Since we mock most of those tests I think trying it out will be most useful

**Comment by alex-l-kong** [comment]
> Looks good. Did you already test out the R code? Since we mock most of those tests I think trying it out will be most useful

Yup, I've tested out the R code already, it works.

File: ark/phenotyping/som_utils_test.py [hunk 1]
```diff
@@ -853,9 +899,9 @@ def test_create_c2pc_data():
         cell_table_path = os.path.join(temp_dir, 'cell_table_size_normalized.csv')
         cell_table.to_csv(cell_table_path, index=False)
 
-        # define a consensus clustered directory
-        pixel_consensus_path = os.path.join(temp_dir, 'pixel_consensus_path')
-        os.mkdir(pixel_consensus_path)
+        # define a pixel data directory
+        pixel_data_path = os.path.join(temp_dir, 'pixel_data_path')
```
[the remark this answers is not in the document]
Comment (alex-l-kong) [reply ark/phenotyping/som_utils_test.py:903]: I've made the change from `pixel_consensus_path` to `pixel_data_path` for this test (`test_create_c2pc_matrix`), however, I need to make the same update for `test_compute_p2c_weighted_channel_avg`.

**Review by cliu72 — APPROVED** [review]
Looks good to me. Tested it a bit on the HIV LN dataset.

**Review by ngreenwald — APPROVED** [review]

**Pull request merged**